### PR TITLE
feat!: ESM + Node 18

### PR DESCRIPTION
**Pull request by mdonnalley**
- Migrate to ESM
- Node 18 minimum
- Repo clean up (prettier, linting, etc...)
- Adds integration test

@W-14259714@

Closes https://github.com/oclif/plugin-update/pull/610

### CODE REVIEW

File: test/integration/sf.integration.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  let stableVersion: string
+  let sf: string
+
+  const versionToUpdateTo = '2.12.7-esm.0'
+  const channel = 'nightly'
+
+  const tarball =
+    process.platform === 'win32'
+      ? `https://developer.salesforce.com/media/salesforce-cli/sf/channels/${channel}/sf-win32-x64.tar.gz`
+      : `https://developer.salesforce.com/media/salesforce-cli/sf/channels/${channel}/sf-linux-x64.tar.gz`
```
Comment (mshanemc) [inline test/integration/sf.integration.ts:80]:
```suggestion
  const platform = process.platform === 'win32' ? 'win32' : 'linux'
  const tarball = `https://developer.salesforce.com/media/salesforce-cli/sf/channels/${channel}/sf-${platform}-x64.tar.gz`
```

File: test/integration/sf.integration.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    const versionResult = await exec(`${sf} version --json`)
+    console.log(versionResult.stdout)
+    expect(versionResult.code).to.equal(0)
+    initialVersion = JSON.parse(versionResult.stdout).cliVersion.replace('@salesforce/cli/', '')
+    console.log('Success!')
+
+    console.log('Linking plugin-update...')
+    // Running `plugins link` is very slow on github-action's windows runners. Writing this file
```
Comment (mshanemc) [inline test/integration/sf.integration.ts:145]: will you change this since https://github.com/oclif/plugin-plugins/pull/679 merged?

**Comment by mshanemc** [comment]
QA notes:

using _@salesforce/cli/2.13.6 darwin-x64 node-v18.15.0_ which is ESM, plus this plugin via `plugins:link`

❌  every command warns about the linked plugin.  Seems fine, unless you run a command with `--json` at which point it's still showing in the terminal output instead of in json result.warnings. 
```shell
 ›   Warning: @oclif/plugin-update is a linked ESM module and cannot be auto-transpiled. Existing compiled source will 
 ›   be used instead.
```

✅  `sf update` (no-op), then run some commands with that version
- `sf org list` 
- `sf org open -o [org]`
- `sf data query -t -q "SELECT MemberType, MemberName, IsNameObsolete, RevisionCounter FROM SourceMember" -o[org]`
- `sf data query -t -q "SELECT MemberType, MemberName, IsNameObsolete, RevisionCounter FROM SourceMember" -o[org]  --json` [this is where I found the json problem for linked esm plugins]

❌  `sf update stable` => `@salesforce/cli: Updating CLI from 2.13.6-9ea464d to 2.11.8-c9872b8`
Should I interpret this as, "you can't update the CLI below the ESM line?"
```
(node:22163) [ERR_UNKNOWN_FILE_EXTENSION] TypeError Plugin: @oclif/plugin-update [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/shane.mclaughlin/eng/oclif-repos/plugin-update/src/commands/update.ts
module: @oclif/core@2.15.0
task: toCached
plugin: @oclif/plugin-update
root: /Users/shane.mclaughlin/eng/oclif-repos/plugin-update
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
```
✅ release notes happen as expected
❌  `sf org list` works, but with the same `ERR_UNKNOWN_FILE_EXTENSION` in the terminal output (it goes to stderr)

❌ now the user can't update back across the line to the esm version (because the update command isn't available)
➜  plugin-update git:(sm/qa-kit-esm) ✗ sf update nightly            
(node:22764) [ERR_UNKNOWN_FILE_EXTENSION] TypeError Plugin: @oclif/plugin-update [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/shane.mclaughlin/eng/oclif-repos/plugin-update/src/commands/update.ts
module: @oclif/core@2.15.0
task: toCached
plugin: @oclif/plugin-update
root: /Users/shane.mclaughlin/eng/oclif-repos/plugin-update
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
 ›   Warning: update nightly is not a sf command.
Did you mean data query? [y/n]: ^C

**Comment by mdonnalley** [comment]
> ❌ every command warns about the linked plugin. Seems fine, unless you run a command with --json at which point it's still showing in the terminal output instead of in json result.warnings.

Confirmed that warning goes to stderr:
```
~/repos/oclif/plugin-update (mdonnalley/esm*) » sf version
 ›   Warning: @oclif/plugin-update is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.
@salesforce/cli/2.12.7-esm.0 darwin-arm64 node-v18.15.0

// silence stderr
~/repos/oclif/plugin-update (mdonnalley/esm*) » sf version 2> /dev/null
@salesforce/cli/2.12.7-esm.0 darwin-arm64 node-v18.15.0

// silence stdout
~/repos/oclif/plugin-update (mdonnalley/esm*) » sf version 1> /dev/null
 ›   Warning: @oclif/plugin-update is a linked ESM module and cannot be auto-transpiled. Existing compiled source will be used
 ›   instead.
```

File: src/hooks/init.ts [hunk 3]
```diff
@@ -1,73 +1,76 @@
...
       const m = await mtime(autoupdatefile)
       let days = 1
       if (opts.config.channel === 'stable') days = 14
-      m.setHours(m.getHours() + (days * 24))
+      m.setHours(m.getHours() + days * 24)
       return m < new Date()
     } catch (error: any) {
-      if (error.code !== 'ENOENT') ux.error(error.stack)
+      if (error.code !== 'ENOENT') throwError(error.stack)
       if ((global as any).testing) return false
```
[the remark this answers is not in the document]
Comment (mdonnalley) [reply src/hooks/init.ts]: I'll just remove it - I'm assuming it was added at some point for testing but I don't see anything setting it anymore

File: test/integration/sf.integration.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+describe('sf integration', () => {
+  let testDir: string
+  let dataDir: string
+  let cacheDir: string
+  let configDir: string
+  let initialVersion: string
+  let stableVersion: string
+  let sf: string
+
+  const versionToUpdateTo = '2.12.7-esm.0'
```
[the remark this answers is not in the document]
Comment (mdonnalley) [reply test/integration/sf.integration.ts:74]: I'll update it to always grab the version just before the current nightly, which will likely always exist

**Review by mshanemc — APPROVED** [review]

**Comment by mshanemc** [comment]
since linked plugins are sketchy, QA round 2 will use an installed qa release `[3.2.4-qa.0](https://github.com/oclif/plugin-update/actions/runs/6497135052)

`sf update nightly`
```
@salesforce/cli: Updating CLI from 2.13.8-01c9e90 to 2.14.0-6c406ae (nightly)... ⣾ 58.3 MB/58.8 MB
```

`sf update stable` (down below ESM line) `2.14.0-6c406ae to 2.12.9-007e29c`
✅ works.  shows RN as expected

`sf update available` does what you'd expect.  That list is getting pretty large

downgrade a long way via --interactive
```
? Select a version to update to 2.2.0
@salesforce/cli: Updating CLI from 2.12.9-007e29c to 2.2.0-4aecedc... ⡿ 20.1 MB/63.1 MB
```
the old `org list` table is so ugly.  Let's get back to nightly

✅ looks great.  🚢

**Pull request merged**